File: CS/Form1.cs
using DevExpress.XtraBars.Ribbon;
using DevExpress.XtraRichEdit;
using DevExpress.XtraRichEdit.API.Native;
using DevExpress.XtraTab;
using System;
using System.Data;
using System.Globalization;
using System.IO;

namespace RichEditMasterDetailMailMerge
{
    public partial class Form1 : RibbonForm
    {
        DataSet xmlDataSet;
        CultureInfo cultureInfo;
        private bool introductionHandled; // flag to ensure Introduction is processed only once
        private RichEditDocumentServer? introductionContent; // cached content

        public Form1()
        {
            InitializeComponent();
            cultureInfo = CultureInfo.CreateSpecificCulture("en-US");

            xtraTabPage1.Tag = mainRichEdit;
            xtraTabPage2.Tag = resultRichEdit;

            mainRichEdit.LoadDocument("Data//template.docx");

            xmlDataSet = new DataSet();
            xmlDataSet.ReadXml("nwind.xml");
            xmlDataSet.Tables["Categories"].PrimaryKey = new DataColumn[] { xmlDataSet.Tables["Categories"].Columns["CategoryID"] };
        }

        private void barButtonItem1_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e)
        {
            MailMergeOptions mailMergeOptions = mainRichEdit.CreateMailMergeOptions();
            mailMergeOptions.DataSource = xmlDataSet;
            mailMergeOptions.DataMember = "Categories";
            mailMergeOptions.MergeMode = MergeMode.NewSection;
            mailMergeOptions.LastRecordIndex = 10;
            resultRichEdit.CalculateDocumentVariable += detail_CalculateDocumentVariable;
            mainRichEdit.MailMerge(mailMergeOptions, resultRichEdit.Document);


            xtraTabControl1.SelectedTabPage = xtraTabPage2;
        }

        void detail_CalculateDocumentVariable(object sender, CalculateDocumentVariableEventArgs e)
        {
            switch (e.VariableName)
            {
                case "ItemsNumber":
                    int itemCount = xmlDataSet.Tables["Products"].Select("CategoryID=" + GetID(e.Arguments[0].Value)).Length;
                    e.Value = itemCount;
                    e.Handled = true;
                    break;
                case "Picture":
                    DataRow? row = xmlDataSet.Tables["Categories"].Rows.Find(GetID(e.Arguments[0].Value));
                    byte[] imageBytes = row?["Picture"] as byte[];

                    if (imageBytes != null)
                    {
                        var imageProcessor = new RichEditDocumentServer();

                        Shape image = imageProcessor.Document.Shapes.InsertPicture(
                            imageProcessor.Document.Range.Start,
                            DocumentImageSource.FromStream(new MemoryStream(imageBytes))
                        );
                        image.TextWrapping = TextWrappingType.InLineWithText;

                        // Set the RichEditDocumentServer with prepared content as the field value.
                        e.Value = imageProcessor;
                        e.Handled = true;
                    }
                    else
                    {
                        // No image found.
                        e.Value = null;
                    }
                    break;
                case "Introduction":
                    // Only handle once. Subsequent occurrences will produce empty content.
                    if (!introductionHandled)
                    {
                        introductionContent = new RichEditDocumentServer();
                        introductionContent.Document.AppendDocumentContent("Data//template-intro.rtf");
                        e.Value = introductionContent;
                        e.Handled = true;
                        introductionHandled = true;
                    }
                    else
                    {
                        // Return empty string (or could skip handling). We handle to suppress further processing.
                        e.Value = DocVariableValue.Current;
                        e.Handled = true;
                    }
                    break;
            }

        }

        protected internal virtual int GetID(string value)
        {
            int result;
            if (Int32.TryParse(value, out result))
                return result;
            return -1;
        }

        void tabControl_SelectedPageChanged(object sender, TabPageChangedEventArgs e)
        {
            RichEditControl richEditControl = (RichEditControl)xtraTabControl1.SelectedTabPage.Tag;
            richEditBarController1.RichEditControl = richEditControl;

        }

    }
}

File: VB/Form1.vb
Imports System.Data
Imports System.Globalization
Imports System.IO
Imports DevExpress.XtraBars.Ribbon
Imports DevExpress.XtraRichEdit
Imports DevExpress.XtraRichEdit.API.Native
Imports DevExpress.XtraTab

Namespace RichEditMasterDetailMailMerge
    Partial Public Class Form1
        Inherits RibbonForm

        Private xmlDataSet As DataSet
        Private cultureInfo As CultureInfo
        Private introductionHandled As Boolean
        Private introductionContent As RichEditDocumentServer

        Public Sub New()
            InitializeComponent()

            xtraTabPage1.Tag = mainRichEdit
            xtraTabPage2.Tag = resultRichEdit

            mainRichEdit.LoadDocument(Path.Combine("Templates", "template.docx"))

            xmlDataSet = New DataSet()
            xmlDataSet.ReadXml("nwind.xml")
            xmlDataSet.Tables("Categories").PrimaryKey =
                New DataColumn() {xmlDataSet.Tables("Categories").Columns("CategoryID")}
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
        End Sub

        Private Sub barButtonItem1_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles barButtonItem1.ItemClick
            ' Reset per-merge state
            introductionHandled = False
            introductionContent = Nothing

            Dim mailMergeOptions As MailMergeOptions = mainRichEdit.CreateMailMergeOptions()
            mailMergeOptions.DataSource = xmlDataSet
            mailMergeOptions.DataMember = "Categories"
            mailMergeOptions.MergeMode = MergeMode.NewSection
            mailMergeOptions.LastRecordIndex = 10

            AddHandler resultRichEdit.CalculateDocumentVariable, AddressOf ResultRichEdit_CalculateDocumentVariable

            mainRichEdit.MailMerge(mailMergeOptions, resultRichEdit.Document)
            xtraTabControl1.SelectedTabPage = xtraTabPage2
        End Sub

        Private Sub ResultRichEdit_CalculateDocumentVariable(ByVal sender As Object, ByVal e As CalculateDocumentVariableEventArgs)
            Select Case e.VariableName
                Case "ItemsNumber"
                    Dim itemCount = xmlDataSet.Tables("Products").Select("CategoryID=" & GetID(e.Arguments(0).Value)).Length
                    e.Value = itemCount
                    e.Handled = True

                Case "Picture"
                    Dim categoryId = GetID(e.Arguments(0).Value)
                    Dim row = xmlDataSet.Tables("Categories").Rows.Find(categoryId)
                    Dim imageBytes As Byte() = Nothing
                    If row IsNot Nothing Then
                        imageBytes = TryCast(row("Picture"), Byte())
                    End If

                    If imageBytes IsNot Nothing AndAlso imageBytes.Length > 0 Then
                        Dim imageProcessor As New RichEditDocumentServer
                        Dim image = imageProcessor.Document.Shapes.InsertPicture(
                            imageProcessor.Document.Range.Start,
                            DocumentImageSource.FromStream(New MemoryStream(imageBytes)))
                        image.TextWrapping = TextWrappingType.InLineWithText
                        e.Value = imageProcessor
                    Else
                        e.Value = "" ' Empty replacement
                    End If
                    e.Handled = True

                Case "Introduction"
                    If Not introductionHandled Then
                        introductionContent = New RichEditDocumentServer
                        introductionContent.Document.AppendDocumentContent(Path.Combine("Templates", "template-intro.rtf"))
                        e.Value = introductionContent
                        introductionHandled = True
                    Else
                        e.Value = "" ' Suppress subsequent occurrences
                    End If
                    e.Handled = True
            End Select
        End Sub

        Protected Friend Overridable Function GetID(ByVal value As String) As Integer
            Dim result As Integer
            If Int32.TryParse(value, result) Then
                Return result
            End If
            Return -1
        End Function

        Private Sub tabControl_SelectedPageChanged(ByVal sender As Object, ByVal e As TabPageChangedEventArgs) Handles xtraTabControl1.SelectedPageChanged
            Dim richEditControl As RichEditControl = CType(xtraTabControl1.SelectedTabPage.Tag, RichEditControl)
            richEditBarController1.RichEditControl = richEditControl
        End Sub
    End Class
End Namespace